### PR TITLE
Copy a book from other pages

### LINKED ISSUE #185: Copy button but from book details or the search view
It makes sense that you can only paste on the actual form, but it would be nice if you could copy from anywhere.

The _easiest_ way to do this is to make a request to get the edit page of the book, and then scrape from that.

### PR DESCRIPTION
**Pull request by braxtonhall**
well that was easy

Kapture 2023-01-05 at 22 21 10

resolves #185 

- [x] On the search page
- [x] On the edition page
- [x] On the work page
- [x] On the edit page

### CODE REVIEW

File: src/manifest.json [hunk 1]
```diff
@@ -11,6 +11,7 @@
 	"content_scripts": [
 		{
 			"matches": ["https://*.librarything.com/*"],
+			"all_frames": true,
```
Comment (adirar111) [inline src/manifest.json:14]: what dis

File: src/ts/content/entities/bookForm/render.ts [hunk 2]
```diff
@@ -36,20 +37,27 @@ const offFormRender = (callback: FormRenderListener): void => {
 const onceFormRender = (callback: FormRenderListener): void =>
 	window.addEventListener(FORM_RENDER_EVENT, encloseCallbackArguments(callback), {once: true});
 
+const onFormRemoved = (callback: () => void): void => window.addEventListener(FORM_REMOVED_EVENT, callback);
+const offFormRemoved = (callback: () => void): void => window.removeEventListener(FORM_REMOVED_EVENT, callback);
+const onceFormRemoved = (callback: () => void): void =>
+	window.addEventListener(FORM_REMOVED_EVENT, callback, {once: true});
+
```
Comment (adirar111) [inline src/ts/content/entities/bookForm/render.ts:44]: very nice

File: src/ts/content/entities/bookForm/util.ts [hunk 3]
```diff
@@ -6,15 +6,15 @@ const getElementsByTag = (parent: HTMLElement) => (tag: string) => Array.from(pa
 
 const getElementsByTags = (parent: HTMLElement, tags: string[]) => tags.flatMap(getElementsByTag(parent));
 
-const getFormElements = (): FormAreaElement[] =>
-	getElementsByTags(getForm(), FORM_DATA_ELEMENT_TAGS) as FormAreaElement[];
+const getFormElements = (document: Document): FormAreaElement[] =>
```
Comment (adirar111) [inline src/ts/content/entities/bookForm/util.ts:9]: why do these parameterize `document`

File: src/ts/content/extensions/copy/onView.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
+import {appendRow, saveFormData} from "./common";
+import * as browser from "webextension-polyfill";
+import {loaderOverlaid} from "../../../common/ui/loadingIndicator";
+import {getDocument} from "../../services/finder/util/getDocument";
+import {getFormData, onFormRemoved} from "../../entities/bookForm";
+
+const handleClick = (href: string) => () =>
+	loaderOverlaid(async () => {
+		const document = await getDocument(href);
```
Comment (adirar111) [inline src/ts/content/extensions/copy/onView.ts:9]: OH

ok this is sick

**Review by adirar111 — APPROVED** [review]
this is so cool

File: src/manifest.json (hunk 1 above)
Comment (braxtonhall) [reply src/manifest.json:14]: UM ok so the main reason is, the search results in LibraryThing are in an iframe. This just makes our stuff apply to iframes as well.

**Pull request merged**